storagemarket/impl/client.go

```diff
@@ -88,93 +75,24 @@ func NewClient(
 		discovery:    discovery,
 		node:         scn,
 
-		deals: deals,
 		conns: map[cid.Cid]network.StorageDealStream{},
-
-		incoming: make(chan *ClientDeal, 16),
-		updated:  make(chan clientDealUpdate, 16),
-
-		stop:    make(chan struct{}),
-		stopped: make(chan struct{}),
-	}
-
-	return c
-}
-
-func (c *Client) Run(ctx context.Context) {
-	go func() {
-		defer close(c.stopped)
-
-		for {
-			select {
-			case deal := <-c.incoming:
-				c.onIncoming(deal)
-			case update := <-c.updated:
-				c.onUpdated(ctx, update)
-			case <-c.stop:
-				return
-			}
-		}
-	}()
-}
-
-func (c *Client) onIncoming(deal *ClientDeal) {
-	log.Info("incoming deal")
-
-	if _, ok := c.conns[deal.ProposalCid]; ok {
-		log.Errorf("tracking deal connection: already tracking connection for deal %s", deal.ProposalCid)
-		return
-	}
-	c.conns[deal.ProposalCid] = deal.s
-
-	if err := c.deals.Begin(deal.ProposalCid, deal); err != nil {
-		// We may have re-sent the proposal
-		log.Errorf("deal tracking failed: %s", err)
-		c.failDeal(deal.ProposalCid, err)
-		return
 	}
 
-	go func() {
-		c.updated <- clientDealUpdate{
-			newState: storagemarket.StorageDealUnknown,
-			id:       deal.ProposalCid,
-			err:      nil,
-		}
-	}()
-}
-
-func (c *Client) onUpdated(ctx context.Context, update clientDealUpdate) {
-	log.Infof("Client deal %s updated state to %s", update.id, storagemarket.DealStates[update.newState])
-	var deal ClientDeal
-	err := c.deals.Get(update.id).Mutate(func(d *ClientDeal) error {
-		d.State = update.newState
-		if update.mut != nil {
-			update.mut(d)
-		}
-		deal = *d
-		return nil
+	statemachines, err := fsm.New(ds, fsm.Parameters{
+		Environment:     c,
+		StateType:       storagemarket.ClientDeal{},
+		StateKeyField:   "State",
+		Events:          clientstates.ClientEvents,
+		StateEntryFuncs: clientstates.ClientStateEntryFuncs,
 	})
-	if update.err != nil {
-		log.Errorf("deal %s failed: %s", update.id, update.err)
-		c.failDeal(update.id, update.err)
-		return
-	}
 	if err != nil {
-		c.failDeal(update.id, err)
-		return
+		return nil, err
 	}
+	c.statemachines = statemachines
+	return c, nil
+}
 
-	switch update.newState {
-	case storagemarket.StorageDealUnknown: // new
-		c.handle(ctx, deal, c.new, storagemarket.StorageDealProposalAccepted)
-	case storagemarket.StorageDealProposalAccepted:
-		c.handle(ctx, deal, c.accepted, storagemarket.StorageDealStaged)
-	case storagemarket.StorageDealStaged:
-		c.handle(ctx, deal, c.staged, storagemarket.StorageDealSealing)
-	case storagemarket.StorageDealSealing:
-		c.handle(ctx, deal, c.sealing, storagemarket.StorageDealNoUpdate)
-		// TODO: StorageDealActive -> watch for faults, expiration, etc.
-	}
+func (c *Client) Run(ctx context.Context) {
```

Comment:
I guess this can boot up in-progress deals when restarting the node?

Reply:
yes, "in the future" :)